ca/ca.go

```diff
@@ -224,35 +224,39 @@ func (ca *CAImpl) CompleteOrder(order *core.Order) {
 	// Check the authorizations - this is done by the VA before calling
 	// CompleteOrder but we do it again for robustness sake.
 	for _, authz := range order.AuthorizationObjects {
+		// Lock the authorization for reading
+		authz.RLock()
 		if authz.Status != acme.StatusValid {
 			return
 		}
+		authz.RUnlock()
 	}
 
+	order.RLock()
 	if order.Status != acme.StatusPending {
-		ca.log.Printf("Error: Asked to complete orrder %s is not status pending, was status %s",
+		ca.log.Printf("Error: Asked to complete order %s is not status pending, was status %s",
 			order.ID, order.Status)
 		return
 	}
+	order.RUnlock()
 
-	ca.log.Printf("Order %s is fully authorized. Ready to issue", order.ID)
 	// Update the order to reflect that we're now processing it
+	order.Lock()
+	defer order.Unlock()
+	ca.log.Printf("Order %s is fully authorized. Ready to issue", order.ID)
 	order.Status = acme.StatusProcessing
 
 	csr := order.ParsedCSR
-	domains := make([]string, len(csr.DNSNames))
-	copy(domains, csr.DNSNames)
-
 	// issue a certificate for the csr
-	cert, err := ca.newCertificate(domains, csr.PublicKey)
+	cert, err := ca.newCertificate(csr.DNSNames, csr.PublicKey)
 	if err != nil {
 		ca.log.Printf("Error: unable to issue order: %s", err.Error())
 		return
 	}
-	ca.log.Printf("Issued certificate serial %s\n", cert.ID)
+	ca.log.Printf("Issued certificate serial %s for order %s\n", cert.ID, order.ID)
 
-	// Update the order to valid status with a certificate URL
+	// Update the order to valid status and store a cert ID for the wfe to use to
+	// render the certificate URL for the order
 	order.Status = acme.StatusValid
-	order.Certificate = order.CertPathPrefix + cert.ID
-	ca.log.Printf("Order %s has Certificate %s", order.ID, order.Certificate)
+	order.CertID = cert.ID
```

Comment:
Since this field is only used to check whether a certificate is done, maybe it would be better to create a field that contains the cert DER, since we'll actually need that later. Otherwise, it's a bit confusing why there are two "ID" fields.